crates/randcast-mock-demo/src/contract/coordinator.rs

```diff
@@ -0,0 +1,251 @@
+}
+
+pub trait Internal {
+    /// A registered participant is one whose pubkey's length > 0
+    fn only_allowed(&self, id_address: &str) -> CoordinatorResult<()>;
+
+    /// The DKG starts when startBlock > 0
+    fn only_when_not_started(&self) -> CoordinatorResult<()>;
+}
+
+pub trait MockHelper {
+    fn mine(&mut self, block_number: usize);
+}
+
+impl MockHelper for Coordinator {
+    fn mine(&mut self, block_number: usize) {
+        self.block_height += block_number;
+
+        println!("coordinator block_height: {}", self.block_height);
+    }
+}
+
+impl Internal for Coordinator {
+    fn only_allowed(&self, id_address: &str) -> CoordinatorResult<()> {
+        if !self.participants.iter().any(|x| x == id_address) {
+            return Err(CoordinatorError::NotAllowlisted);
+        }
+
+        Ok(())
+    }
+
+    fn only_when_not_started(&self) -> CoordinatorResult<()> {
+        if self.start_block != 0 {
+            return Err(CoordinatorError::AlreadyStarted);
+        }
+
+        Ok(())
+    }
+}
+
+impl Transactions for Coordinator {
+    fn start(
+        &mut self,
+        block_height: usize,
+        members: HashMap<String, Vec<u8>>,
+    ) -> CoordinatorResult<()> {
+        self.only_when_not_started()?;
+
+        self.start_block = block_height;
+
+        self.block_height = block_height;
+
+        members.into_iter().for_each(|(address, key)| {
+            self.participants.push(address.clone());
+
+            self.keys.insert(address, key);
+        });
+
+        println!("coordinator deployed. block_height: {}", self.block_height);
+
+        Ok(())
+    }
+
+    fn publish(&mut self, id_address: String, value: Vec<u8>) -> CoordinatorResult<()> {
+        self.only_allowed(&id_address)?;
+
+        let blocks_since_start = self.block_height - self.start_block;
+
+        if blocks_since_start <= self.phase_duration {
+            if self.shares.contains_key(&id_address) {
+                return Err(CoordinatorError::SharesExisted);
+            }
+
+            self.shares.insert(id_address, value);
+        } else if blocks_since_start <= 2 * self.phase_duration {
+            if self.responses.contains_key(&id_address) {
+                return Err(CoordinatorError::SharesExisted);
+            }
+
+            self.responses.insert(id_address, value);
+        } else if blocks_since_start <= 3 * self.phase_duration {
+            if self.justifications.contains_key(&id_address) {
+                return Err(CoordinatorError::SharesExisted);
+            }
+
+            self.justifications.insert(id_address, value);
+        } else {
+            return Err(CoordinatorError::DKGEnded);
+        }
+
+        Ok(())
+    }
+}
+
+impl Views for Coordinator {
+    fn get_shares(&self) -> CoordinatorResult<Vec<Vec<u8>>> {
+        Ok(self
+            .participants
+            .iter()
+            .map(|id_address| {
+                self.shares
+                    .get(id_address)
```

Comment:
can try to use [filter_map](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.filter_map)